### DISCUSSION TITLE
Is there any way to rebuild only changed parts of package with `build-package.sh`?

### DISCUSSION THREAD

**Discussion by twaik** [termux/termux-packages#19654] [Q&A] [answered]
Hello. 
I want to patch a big project, so building it with `build-package.sh` takes a while.
And changes are pretty much small so rebuilding the whole project if I change a few lines is not very efficient.
Is there a way to make `build-package.sh` not clean the build directory and simply make `ninja`/`make` and build the deb? Without rebuilding everything?

I know I can use on-device building for this, but this does not seem to be really comfortable to me.
Thank you.

**Comment by Biswa96** [comment]
I miss the good ol' days with `makepkg --noextract --noprepare` 😞

**Comment by agnostic-apollo** [comment, marked as answer]
- https://github.com/termux/termux-packages/commit/ecd214881aa7e8b8ab936dd1f5a7207864e71806
- https://github.com/termux/termux-packages/wiki/Building-packages#build-options

**Reply by twaik** [reply]
How could I miss it...

**Reply by agnostic-apollo** [reply]
Do you really want me to answer that? :p

**Reply by twaik** [reply]
Yes. Tell me I am a worst man ever, not reading docs of script I am using for a long period of time.

**Reply by agnostic-apollo** [reply]
You are the worst man ever, not reading docs of script you have been using for a long period of time.

Actually, I found out myself couple of days ago, but I don't port packages, so never needed to look into it. But there are lot of bugs in `build-package.sh`, so I don't know if it works properly.

**Reply by twaik** [reply]
Probably we should add this to `build-package.sh`'s help text because it is not there.
```

Build a package by creating a .deb file in the debs/ folder.

Available options:
  -a The architecture to build for: aarch64(default), arm, i686, x86_64 or all.
  -d Build with debug symbols.
  -D Build a disabled package in disabled-packages/.
  -f Force build even if package has already been built.
  -F Force build even if package and its dependencies have already been built.
  -i Download and extract dependencies instead of building them.
  -I Download and extract dependencies instead of building them, keep existing /data/data/com.termux/files files.
  -L The package and its dependencies will be based on the same library.
  -q Quiet build.
  -w Install dependencies without version binding.
  -s Skip dependency check.
  -o Specify directory where to put built packages. Default: output/.
  --format Specify package output format (debian, pacman).
  --library Specify library of package (bionic, glibc).
```

**Reply by agnostic-apollo** [reply]
Already did that, will push soon.

**Reply by Biswa96** [reply]
I have added  a pull request with that change sometime ago.

**Reply by twaik** [reply]
For some reason mesa package produced with `./build-package.sh -I -f -c mesa` is not installable.
```
WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Reading package lists...
Building dependency tree...
Reading state information...
Suggested packages:
  mesa-dev
The following packages will be upgraded:
  mesa
1 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
1 not fully installed or removed.
Need to get 0 B/3320 kB of archives.
After this operation, 8192 B of additional disk space will be used.
Get:1 /data/data/com.termux/files/home/mesa_24.0.5_aarch64.deb mesa aarch64 24.0.5 [3320 kB]
(Reading database ... 42146 files and directories currently installed.)
Preparing to unpack .../home/mesa_24.0.5_aarch64.deb ...
Unpacking mesa (24.0.5) over (24.0.5) ...
dpkg: error processing archive /data/data/com.termux/files/home/mesa_24.0.5_aarch64.deb (--unpack):
 error creating directory './DEBIAN': Read-only file system
dpkg: error while cleaning up:
 unable to remove newly-extracted version of '/DEBIAN': Read-only file system
dpkg-deb: error: paste subprocess was killed by signal (Broken pipe)
Errors were encountered while processing:
 /data/data/com.termux/files/home/mesa_24.0.5_aarch64.deb
E: Sub-process /data/data/com.termux/files/usr/bin/dpkg returned an error code (1)
```
Adding `rm -rf ${TERMUX_PKG_MASSAGEDIR}/DEBIAN` to `termux_step_post_make_install` fixed the thing, but that is not an option for merging to master.
